netket/optimizer/qgt/qgt_jacobian_pytree.py

```diff
@@ -39,7 +39,7 @@ def QGTJacobianPyTree(
         )
```

Comment:
I don't know who wrote this, but this won't work: the eval shape gets apply_fun which is most likely an HahablePartial and it's not compatible with jax.eval_shape..

Reply:
I wrote it, I copied it out from somewhere in @inailuig's original autodetection logic without really understanding it (jax magick...)
Can you think of something that would do what I intended to do?

Reply:
also, does it need something along these lines?
```python
    # pre-apply the model state
    def forward_fn(W, σ):
        return apply_fun({"params": W, **model_state}, σ)
```

Reply:
eval shape basically evaluates the shape and dtype of the result (somehow):
```
def eval_shape(fun, *args) :
     res = fun(*args)
     return res.dtype, res.shape
```

It's just that, annoyingly, it does not support out of the box arbitrary callable but just functions.
There are many workarounds to do it. I

Reply:
Do you mean something like this?
```python
def f(W,σ):
    return vstate._apply_fun({"params": W, **model_state}, σ)
complex_output = jax.eval_shape(vstate._apply_fun, vstate.parameters, vstate.samples).dtype
```

Reply:
> also, does it need something along these lines?
> 
> ```python
>     # pre-apply the model state
>     apply_fun({"params": W, **model_state}, σ)
> ```

I wasn't referring to this issue so much as the facility to provide `model_state` (it's done in many places, so I kept copying, but I'm still not sure what it achieves or how you can use it...)

Reply:
Update: this is exactly the error I caught in https://github.com/netket/netket/issues/701#issuecomment-835834164 so I tried these fixes. Turns out `eval_shape` is perfectly happy with `vstate._apply_fun` as it is, but it does need these extra arguments. See a working fix below.